### PR TITLE
Bigtable: refactoring read_rows infrastructure

### PR DESCRIPTION
**Pull request by sduskis**
This work is on behalf of understanding the issues behind issue #5876 where InvalidChunk is thrown due to row keys being processed out of order.
About this PR
- introducing `PartialRowsData._process_chunk` to break up logic in `__iter__`, and allow for additional testability.  `__iter__` is responsible to get chunks, call `_process_chunk`, and manage `last_scanned_row_key`.
- removed `_copy_from_current`
- _state is now explicitly set
- fixed a bug in `PartialRowsData.counter` which was counting chunks instead of rows.  This counter is used in reconstructing requests for "smart retries" that update the request on transient exceptions

### CODE REVIEW

File: bigtable/google/cloud/bigtable/row_data.py [hunk 1]
```diff
@@ -440,86 +422,91 @@ def __iter__(self):
...
-
-            row = self._row
-            cell = self._cell
-
             for chunk in response.chunks:
+                self._process_chunk(chunk)
+                if chunk.commit_row:
+                   self.last_scanned_row_key = self._previous_row.row_key
+                   self._counter += 1
+                   yield self._previous_row
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/row_data.py]: Please add one space to the three lines above to fix lint

File: bigtable/google/cloud/bigtable/row_data.py [hunk 2]
```diff
@@ -440,86 +422,91 @@ def __iter__(self):
...
 
-                if chunk.reset_row:
-                    self._validate_chunk_reset_row(chunk)
-                    row = self._row = None
-                    cell = self._cell = self._previous_cell = None
-                    continue
+            if (response.last_scanned_row_key and
+                response.last_scanned_row_key > self.last_scanned_row_key):
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/row_data.py]: Please add 4 spaces to fix lint

File: bigtable/google/cloud/bigtable/row_data.py [hunk 3]
```diff
@@ -440,86 +422,91 @@ def __iter__(self):
...
 
-                if chunk.commit_row:
-                    if chunk.value_size > 0:
-                        raise InvalidChunk()
+        self._update_cell(chunk)
 
-                    self._save_current_cell()
+        if self._row is None:
+            if (self._previous_row is not None and
+                self._cell.row_key <= self._previous_row.row_key):
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/row_data.py]: Please add 4 spaces to fix lint

File: bigtable/google/cloud/bigtable/row_data.py [hunk 4]
```diff
@@ -440,86 +422,91 @@ def __iter__(self):
...
+            self._validate_cell_data_new_cell()
+        else:
+            self._validate_cell_data_cell_in_progress(chunk)
+            self._cell.append_value(chunk.value)
 
-                    yield self._row
+    def _validate_cell_data_new_cell(self):
+        if (not self._cell.row_key or
+            not self._cell.family_name or
+            self._cell.qualifier is None):
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/row_data.py]: Please add 4 spaces here to fix lint

File: bigtable/google/cloud/bigtable/row_data.py [hunk 5]
```diff
@@ -440,86 +422,91 @@ def __iter__(self):
...
+            raise InvalidChunk()
 
-                    self.last_scanned_row_key = self._row.row_key
-                    self._row, self._previous_row = None, self._row
-                    self._previous_cell = None
-                    row = cell = None
-                    continue
+        if (self._previous_cell and
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/row_data.py]: Please remove trailing space after `and` to fix lint

File: bigtable/google/cloud/bigtable/row_data.py [hunk 6]
```diff
@@ -440,86 +422,91 @@ def __iter__(self):
...
+            self._cell.qualifier is None):
+            raise InvalidChunk()
 
-                    self.last_scanned_row_key = self._row.row_key
-                    self._row, self._previous_row = None, self._row
-                    self._previous_cell = None
-                    row = cell = None
-                    continue
+        if (self._previous_cell and 
+            self._previous_cell.row_key != self._cell.row_key):
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/row_data.py]: Please add 4 spaces here to fix lint

File: bigtable/google/cloud/bigtable/row_data.py [hunk 7]
```diff
@@ -440,86 +422,91 @@ def __iter__(self):
...
-        if self._state == self.STATE_CELL_IN_PROGRESS:
-            self._copy_from_current(cell)
-
-    def _validate_cell_data_new_row(self, cell):
-        if (not cell.row_key or
-                not cell.family_name or
-                cell.qualifier is None):
+    def _validate_cell_data_cell_in_progress(self, chunk):
+        if (chunk.row_key and
+            chunk.row_key != self._cell.row_key):
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/row_data.py]: Please add 4 spaces here to fix lint

File: bigtable/google/cloud/bigtable/row_data.py [hunk 8]
```diff
@@ -440,86 +422,91 @@ def __iter__(self):
...
             raise InvalidChunk()
 
-        if (self._previous_row is not None and
-                cell.row_key <= self._previous_row.row_key):
+        if (chunk.HasField('family_name') and
+            chunk.family_name.value != self._cell.family_name):
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/row_data.py]: Another 4 spaces here please

File: bigtable/google/cloud/bigtable/row_data.py [hunk 9]
```diff
@@ -440,86 +422,91 @@ def __iter__(self):
...
             raise InvalidChunk()
 
-    def _validate_cell_data_row_in_progress(self, cell):
-        if ((cell.row_key and
-             cell.row_key != self._row.row_key) or
-                (cell.family_name and cell.qualifier is None)):
+        if (chunk.HasField('qualifier') and
+            chunk.qualifier.value != self._cell.qualifier):
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/row_data.py]: Another 4 spaces here please

File: bigtable/google/cloud/bigtable/row_data.py [hunk 10]
```diff
@@ -544,32 +532,17 @@ def _save_current_cell(self):
...
             if not cell.row_key:
                 cell.row_key = previous.row_key
-            if not cell.family_name:
-                cell.family_name = previous.family_name
-            # NOTE: ``cell.qualifier`` **can** be empty string.
-            if cell.qualifier is None:
-                cell.qualifier = previous.qualifier
+                if not cell.family_name:
+                    cell.family_name = previous.family_name
+                    NOTE: ``cell.qualifier`` **can** be empty string.
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/row_data.py]: Please comment out this line back

File: bigtable/tests/unit/test_row_data.py [hunk 11]
```diff
@@ -439,15 +439,60 @@ def test_state_new_row_w_row(self):
...
+        yrd._response_iterator = iterator
+        rows = [row for row in yrd]
+        result = rows[0]
+        self.assertEqual(result.row_key, self.ROW_KEY)
+        self.assertEqual(yrd._counter, 1)
+
+        yrd._row = object()
+        self.assertEqual(yrd.state, yrd.NEW_ROW)
+
+    def test_multiple_chunks(self):
```
Comment (AVaksman) [inline bigtable/tests/unit/test_row_data.py]: This should be removed as it is a duplicate from [line 451](https://github.com/GoogleCloudPlatform/google-cloud-python/pull/5956/files#diff-e1d4ea33ffcf7ba98a70d7dc9af8ec6eR451)

**Comment by sduskis** [comment]
@tseaver, I ran `git push upstream minor_read_changes`.  Fingers crossed...

**Comment by sduskis** [comment]
@tseaver, I'll open a new PR from the local branch.

**Pull request closed without merging**